Add basic documentation on using Snippets to DocC documentation

heckj opened the pull request:
Work towards resolving #417

## Summary

Adds some base line documentation that describes how to use the snippets feature built in to DocC and the DocC Plugin.

- how to add snippets to your project
- how to run and verify the snippets
- how to make slices in a snippet
- how to reference the snippet, or a slice from a snippet, from within another DocC catalog entry

## Dependencies & Testing

- no dependencies
- visually inspected with a local preview build of the DocC documentation for testing

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [ ] Added tests (no tests, doc updates only)
- [X] Ran the `./bin/test` script and it succeeded
- [X] Updated documentation if necessary

(this pull request is the combined work of https://github.com/JacobHearst and myself, I just happen to be submitting it)

d-ronnqvist commented:
> (this pull request is the combined work of https://github.com/JacobHearst and myself, I just happen to be submitting it)

FYI: You can coauthor commits by including a line like this in the commit message:

```
Co-authored-by: Jacob Hearst <that-persons@email.address>
```

d-ronnqvist commented on Sources/docc/DocCDocumentation.docc/Reference Syntax/API Reference Syntax/Snippet.md:
```diff
@@ -0,0 +1,26 @@
+# ``docc/Snippet``
```
This page won't be displayed as long as the snippet symbol isn't in the symbol graph file of directives that this catalog uses.

To include it you'd need to remove [this `hiddenFromDocumentation = true` line](https://github.com/swiftlang/swift-docc/blob/main/Sources/SwiftDocC/Semantics/Snippets/Snippet.swift#L33) and then run `swift run generate-symbol-graph` in the DocC repo and commit the modified symbol graph file.

heckj replied:
Thank you! I don't know how long it would have taken me to hunt down that detail without your guidance!

d-ronnqvist commented on Sources/docc/DocCDocumentation.docc/snippets.md:
```diff
@@ -0,0 +1,144 @@
+
+## Overview
+
+...tbd...
+- Describe the problem and summarize the developer action.
+- Explain why the problem is relevant and provide context for the task. Don't simply repeat the abstract; expand on it.
+- Keep the Overview to one or two paragraphs.
+- For very short articles that consist of just a couple of paragraphs, all of the content can be in the Overview.
+
+...tbd...
```
It might be good to assume that the reader doesn't know the destination between a code snippet and a markdown code block element and lead with that information. A reader who finds this information might be overwhelmed of all the steps if all they want to do is to write a code block in their markup. That distinction might serve as a good segue to the benefits of compiled code examples which lets the reader make an informed decision about using code blocks or code snippets in their documentation.

heckj commented:
@JacobHearst @d-ronnqvist sorry for the delay in writing an overview. Added one based on @d-ronnqvist feedback, see what you think.

heckj commented:
@swift-ci please test

heckj commented:
@d-ronnqvist the test failure is asserting a missing copyright header, which is included - but I suspect the check (which I can't see, since it's internal CI) isn't allowing `2025` as a year, or a stand-alone year. From visual inspection, it seems to match what's expected.

d-ronnqvist commented:
> @d-ronnqvist the test failure is asserting a missing copyright header, which is included - but I suspect the check (which I can't see, since it's internal CI) isn't allowing `2025` as a year, or a stand-alone year. From visual inspection, it seems to match what's expected.

The check is defined in [check-source (here)](https://github.com/swiftlang/swift-docc/blob/main/bin/check-source#L21) with the specific check for markdown files [later in that file](https://github.com/swiftlang/swift-docc/blob/main/bin/check-source#L79-L85).

heckj commented:
@swift-ci please tests

patshaughnessy commented on Sources/docc/DocCDocumentation.docc/DocC Documentation.md:
```diff
@@ -25,6 +25,7 @@ DocC syntax — called _documentation markup_ — is a custom variant of Markdow
 - <doc:writing-symbol-documentation-in-your-source-files>
 - <doc:adding-supplemental-content-to-a-documentation-catalog>
 - <doc:linking-to-symbols-and-other-content>
+- <doc:snippets>
```
```suggestion
- <doc:adding-code-snippets-to-your-content>
```

I'd rename the `snippets.md` with your new article to make it more obvious that this is a link to an article.

patshaughnessy commented on Sources/docc/DocCDocumentation.docc/snippets.md:
```diff
@@ -0,0 +1,142 @@
+# Adding Code Snippets to your Content
```
As mentioned above, I'd suggest using a longer file name for this markdown article, to make it more obvious the link above is to the article and not the symbol (the directive).

Maybe: `adding-code-snippets-to-your-content.md`

patshaughnessy commented on Sources/docc/DocCDocumentation.docc/snippets.md:
```diff
@@ -0,0 +1,142 @@
+
+DocC parses contiguous comments within your the code of a snippet as markdown to annotate your code when embedded in documentation.
+DocC will attempt to reference symbols from within these comments just like any other documentation content.
+You can reference symbols from your API, which DocC converts into hyperlinks to that symbol when displaying the content.
```
```suggestion
You can reference symbols from your API, which DocC converts into hyperlinks to that symbol when displaying the content.

## Topics

### Directives

- ``Snippet``
```

You should curate (link) to the directive symbol from here or from somewhere else; otherwise the new `Snippet` directive page will appear as a link to an uncurated symbol from the root page under a section called "Classes."

patshaughnessy commented on Sources/docc/DocCDocumentation.docc/Reference Syntax/API Reference Syntax/Snippet.md:
```diff
@@ -0,0 +1,25 @@
+- Parameters:
+    - path: A reference to the location of a code example.
+    - slice: The name of a section within the code example that you annotate with comments in the snippet. **(optional)**
+
+## Overview
```
```suggestion
```

Remove `## Overview` from here or we end up with "Overview" appearing twice on the page.

patshaughnessy reviewed (changes requested):
This is a great! Thank you so much for writing these docs 👏  This is a very cool feature that most people aren't even aware of.

Just a curation issue and another minor tweak please.

heckj commented:
Thanks @patshaughnessy - accepted all suggestions and added a commit to rename the file per your suggestion.

heckj commented:
@swift-ci please test

patshaughnessy commented:
Hello @heckj  - sorry for the delay.

I wasn't able to make this work following the instructions in your new article. I suspect that Swift DocC users have to add the Swift DocC Plugin as a dependency in their Package.swift, and then build the docs using the `swift package generate-documentation` CLI command. (Is there a way to use this feature from Xcode without SPM?)

Would you mind updating the article to show how to do this?

For example, the article might include details about these steps:
- Create an example SPM project (e.g. "Hello World")
- Add `https://github.com/apple/swift-docc-plugin` as a dependency
- Create a snippet (you have this already)
- Reference the snippet using the `@Snippet` directive (you have this already)
- Build the docs using the Swift DocC Plugin (`swift package generate-documentation`)
- View the snippet by opening a Ruby/Python web server from the subdirectory the plugin displays.

Without all of this information spelled out, I don't know how many readers will be able to use the feature successfully.

Let me know if you need help and I can push a draft of what this might look like.

heckj commented:
Thanks for looking @patshaughnessy !

I don't think you strictly need the plugin to do this, but it's certainly easier than wrangling the independent steps of using `docc` on the command-line from the toolchain.  I can (and will) revise to include your suggested path as a walk-through to make it easier for someone to use the features described.

To my knowledge, this capability hasn't been enabled with Xcode, and is only functional when building and rendering documentation using the Swift Package Manager. The core pieces for building the snippets are part of that project, and generally I do tend to add the swift-docc-plugin to a package when I'm working on such documentation.

heckj commented:
@patshaughnessy I added in an early section explicitly calling out adding the swift-docc-plugin in order to view the snippets, but stopped short of adding sections that illustrate creating a project and making this more of a tutorial-like walkthrough, since that's not done for any of the other proximate documentation.

Give it a read and see what you think, and I'd happily take updates or a draft of what you'd like to see to make this easier to use. (I've also rebased this branch to catch it up a bit)

heckj commented:
@swift-ci please test

heckj commented:
Looks great @patshaughnessy - Thanks for the additions!

No worries with the comments in snippets as markdown - happy to leave that to follow ups and just get the baseline docs out there!

d-ronnqvist commented on Sources/docc/DocCDocumentation.docc/adding-code-snippets-to-your-content.md:
```diff
@@ -0,0 +1,213 @@
+# Adding Code Snippets to your Content
+
+@Metadata {
+    @Available("Swift", introduced: "5.7")
```
This is different from the `introducedVersion` on the Snippet type. We should update one of these values so that they are both correct.

heckj replied:
I updated it to 5.7 - while it was technically released with 5.6, it wasn't really operational until after 5.7 from what I saw due to a couple lingering issues. I can flip it back to 5.6 if you'd prefer.

heckj commented:
Thanks David - I'll go back and revise to align properly and shift some of the content  - hadn't clued into `@ DirectiveArgumentWrapped` - thanks for that reference!

heckj commented on Sources/docc/DocCDocumentation.docc/adding-code-snippets-to-your-content.md:
```diff
@@ -0,0 +1,213 @@
+  ├─ Package.swift
+  ├─ Snippets/
+  │  ├─ example-snippet.swift
+  ├─ Sources/
+  │  ├─ YourProject.docc/
+  │  │ ├─ YourProject.md
+  │  ├─ main.swift
```
It's correct, if exceedingly awkward. When you ask SwiftPM to initialize a tool or executable package, it drops in a main.swift directly under Sources, and there's code to allow that sneaky trick on the idea that it's an optimization to make things "easier" - I personally lean towards "just annoyingly inconsistent" - but it IS correct.

A default "library" package initialization would have a directory with the target name and it's source inside that directory.

d-ronnqvist replied:
In that case, can we show the library setup instead? 

Executables don't have Swift API to show examples of but libraries do.

heckj commented on Sources/docc/DocCDocumentation.docc/Reference Syntax/API Reference Syntax/Snippet.md:
I'm not entirely sure where you have in mind to place it. 

I tried dropping it on the the class itself with the following block:
```
/// Embeds a code example from the project's code snippets.
///
/// ```markdown
/// @Snippet(path: "my-package/Snippets/example-snippet", slice: "setup")
/// ```
///
/// Place the `Snippet` directive to embed a code example from the project's snippet directory.
/// The path to the snippet ///is identified with three parts:
///
/// 1. The package name as defined in `Package.swift`
///
/// 2. The directory path to the snippet file, starting with "Snippets".
///
/// 3. The name of your snippet file without the `.swift` extension
///
/// If the snippet had slices annotated within it, an individual slice of the snippet can be referenced with the ///`slice` option.
/// Without the option defined, the directive embeds the entire snippet.
```

But with the sidecar page from the catalog removed, the content didn't appear.

heckj commented on Sources/docc/DocCDocumentation.docc/Reference Syntax/API Reference Syntax/Snippet.md:
```diff
@@ -0,0 +1,23 @@
+
+- Parameters:
+    - path: A reference to the location of a code example.
+    - slice: The name of a section within the code example that you annotate with comments in the snippet. **(optional)**
```
removed the extraneous parameters

d-ronnqvist commented on Sources/SwiftDocC/Semantics/Snippets/Snippet.swift:
```diff
@@ -12,6 +12,27 @@ import Foundation
 public import Markdown
 import SymbolKit
 
+// NOTE: If you update the DocC content, remember to run
+// `swift run generate-symbol-graph` to regenerate the DocC
+// symbol content for the DocC Documentation.
```
I don't think a specific directive file is the right place for this information. 

It's also not just about the documentation comment about the public properties of this class (and their documentation). This also applies to all `AutomaticDirectiveConvertible` types (including the introduction of new types).

If anything, maybe this information belongs in a section of the Directives file in the contributor documentation similar to how the Benchmarking file has a section about adding new metrics and there's an entire article about adding feature flags.

heckj commented:
@swift-ci please test

d-ronnqvist commented on Sources/docc/DocCDocumentation.docc/adding-code-snippets-to-your-content.md:
```diff
@@ -0,0 +1,213 @@
+
+### Documenting the code in your snippet
+
+DocC parses contiguous comments within the code of a snippet as markdown to annotate your code when embedded in documentation.
+DocC will attempt to reference symbols from within these comments just like any other documentation content.
+You can reference symbols from your API, which DocC converts into hyperlinks to that symbol when displaying the content.
```
I don't understand why this conversation was marked as resolved. My understanding is that you said that you meant to remove this section but it's still here.

heckj replied:
I'm sorry David - my mistake entirely. I had thought I had removed it, but clearly I didn't. Fixing in a commit now.

heckj commented:
@swift-ci please test

heckj commented on Sources/docc/DocCDocumentation.docc/adding-code-snippets-to-your-content.md:
````diff
@@ -0,0 +1,213 @@
+
+To generate or preview documentation with snippets, add [swift-docc-plugin](https://github.com/apple/swift-docc-plugin) as a dependency to your package.
+
+For example, use the command:
+
+```bash
+swift package add-dependency https://github.com/apple/swift-docc-plugin --from 1.1.0
+```
+
+Or edit your `Package.swift` to add the dependency:
+
+```
+let package = Package(
+    // name, platforms, products, etc.
+    dependencies: [
+        // other dependencies
+        .package(url: "https://github.com/swiftlang/swift-docc-plugin", from: "1.1.0"),
+    ],
+    targets: [
+        // targets
+    ]
+)
+```
````
Leaving this in place due to Pat's feedback of explicitly wanting to see it.
